Fitch wind farm scheme bug fix: axial induction factor

SengersB opened the pull request:
Bug fix and very small dev in Fitch wind farm parameterization

TYPE: bug fix 

KEYWORDS: fitch, wind farm parameterization, turbine, wind farm, wind energy

SOURCE: Balthazar Sengers, Fraunhofer IWES, Germany

DESCRIPTION OF CHANGES:
Problem:
1. In the standard Fitch wind farm parameterization, the local (grid cell) wind speed is used to calculate power and thrust. This wind speed is however reduced because of induction. Power and thrust curves (look-up tables) are determined using free inflow wind speed.
2. The turbine thrust is a relevant metric that is not outputted. 

Solution:
1. The axial induction factor is calculated and used to correct the local (grid cell) wind speed back to the free inflow wind speed. This is done in 2 iterations.  Code becomes an additional option for 'windfarm_opt=1' called windfarm_induction = 1 (default on) where 0 is old version.
2. The turbine thrust is calculated and passed through so it can be outputted now. 

LIST OF MODIFIED FILES: 
M   Registry/Registry.EM_COMMON
M   dyn_em/module_first_rk_step_part1.F
M   phys/module_pbl_driver.F
M   phys/module_wind_fitch.F
A    doc/licenses/fraunhofer_license.txt

TESTS CONDUCTED: 
1. Yes, demonstrated in paper: https://doi.org/10.5194/wes-9-1689-2024
2. The regression tests have passed.

RELEASE NOTE: 
Fitch wind farm parameterization: correct local (grid cell) wind speed to free wind speed using axial induction correction proposed by Vollmer et al. (2024) https://doi.org/10.5194/wes-9-1689-2024

weiwangncar commented:
@SengersB Thank you very much for your contribution to the WRF model. Your code has many many differences from the develop branch in our repository. For example, your LICENSE.txt file is different from our version. There maybe many more modified files (out of 267 files) are not part of your bug fix. Please clone the latest WRF model code, and add only the code that pertaining to the bug fix to this PR.

SengersB commented:
@weiwangncar Thanks for the quick response. I used the branch `release-v4.7.1` as a base. Is that not correct?

weiwangncar commented:
@SengersB Starting from WRF v4.7.1 is good. But did you really changed 267 files to implement your Fitch fix? It is unlikely. You should identify the files changed in your original code, and add them to the new release. If you started with WRFv4.7.1, the changes that are already in that version should not show up as updates. See this [page](https://github.com/wrf-model/WRF/wiki/Workflow-for-WRF-Code-Modification) for information on how to make code changes.

weiwangncar commented:
@SengersB Is your intention to contribute your modification to our repository? Or do you plan to maintain the code in your own branch?

weiwangncar commented:
@SengersB It still says 'Files changed 278'. See the last item on top of your PR message.

weiwangncar commented:
@SengersB Our master and develop branches are the same. You should first clone the WRF repository code (make sure that code is version 4.7.1 by checking the top level README file), and then add your changes in those five files to the cloned code. Please see the page I provided early for instructions.

Commit:
rebased

dudhia reviewed (dismissed):
code looks safe to add - only adds a diagnostic

SengersB commented:
@islas Hmm I cannot figure out how to give you access. Steps listed [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) don't work: checkbox is simply missing.. Could this be due to restrictions on our side?

islas commented:
@SengersB I'm not sure. I've never encountered that issue. Perhaps we can try this:

Here is a patch file with the edits that I did. You can download the file and apply it to your branch. From there you can commit & push the changes.

[pr_2242_diff.txt](https://github.com/user-attachments/files/25024374/pr_2242_diff.txt)

You can apply using `git apply <filename>`

SengersB commented:
@islas Thanks, I'll make the changes. 

@weiwangncar @dudhia The last days we were advertising to some other heavy users of the wind farm parameterizations that this bug fix will hopefully be in the next WRF release. They expressed the desire to keep the original formulation (without this proposed bugfix) also available. I therefore propose to implement the bugfix as an addition option (default) in windfarm_opt. 
I can code this up, add the reference to the Fraunhofer licence and push these changes. 
Would this be acceptable to do this still within this pull request? My apologies for the additional work this might cause.

dudhia commented:
I enncourage keeping the previous code as an option. I had misread this as just providing a new diagnostic thrust, but now I see it alters results. It can be done within the same PR.

Commit:
induction option added; licensing statements added

SengersB commented:
Hi @dudhia @weiwangncar,

I added the discussed changes and pushed the updated code. In short:
- I added a new namelist parameter 'windfarm_induction'. If turned on (1 = default), the proposed bugfix is activated; if turned off (0) the original code is used
- I added the licensing statement around all modified lines.

Again apologies for the mess with the licensing. Hope everything is fine now and we can merge this bugfix after your checks. 

Thanks, Balthazar

dudhia commented:
OK, we need to check which automated tests failed and get back to you.

weiwangncar commented on Registry/Registry.EM_COMMON:
```diff
@@ -3395,7 +3394,9 @@ package   wrfhydro       wrf_hydro==1                -            state:SOLDRAIN
-package   fitchscheme    windfarm_opt==1             -            state:power
+# THE FOLLOWING LINES OF CODE ARE LICENSED UNDER doc/licenses/fraunhofer_license.txt
+package   fitchscheme    windfarm_opt==1             -            state:power,thrust
+# END LICENSE
```
@SengersB May need a separate package line just for the variable 'thrust', since 'power' is part of the original scheme.

SengersB replied:
Agreed. I don't know the syntax on how to do this here. If this is an issue, I'd be okay with removing the licensing statement here

weiwangncar commented on run/README.namelist:
```diff
@@ -1133,6 +1133,9 @@ Options for wind turbine drag parameterization:
                                                       in terms of lat-lon with the filename of 'windturbines-ll.txt'
+ ! THE FOLLOWING LINES OF CODE ARE LICENSED UNDER doc/licenses/fraunhofer_license.txt                                                    
+ windfarm_induction                 = 1         ! Use the axial induction correction as described by Vollmer et al (2024). Valid only with 'windfarm_opt=1'. 0: Off, 1: On (default)
+ ! END LICENSE
```
@SengersB Do you want to the default option to be the new one?

SengersB replied:
If that's ok with you, that'd be my preference

weiwangncar commented:
@SengersB Your latest changes have some compilation errors:
```
  115 |   REAL :: a_ind
      |               1
Error: Symbol 'a_ind' at (1) already has basic type of REAL
module_wind_fitch.f90:116:17:

  116 |   REAL :: thrust1
      |                 1
Error: Symbol 'thrust1' at (1) already has basic type of REAL
```

Commit:
bugfix module_wind_fitch.F

SengersB commented:
@weiwangncar Apologies, I fixed these compilation errors now.

Unfortunately I cannot run a full test on this code, as v4.7.1 does not want to compile on our cluster and we're currently without IT support. So I had to make these changes in v4.6.1 and transfer them, but apparently missed something during the merging. I apologize again

weiwangncar commented:
@SengersB Thanks for the fix. Now the regression test results look good:

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 23           24
    Number of Builds       : 60           57
    Number of Simulations  : 158           150        0
    Number of Comparisons  : 95           86        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

weiwangncar commented:
@SengersB Can you address my comment for your registry change?

Commit:
fix licensing statement in Registry

SengersB commented:
@weiwangncar I have now removed the licensing statement around this one line

SengersB commented:
Hi @weiwangncar @dudhia ,

I noticed today that I made a mistake in the last version of the code when trying to retain the original Fitch WFP. Actually, part of the new correction was still active, which is not what was intended. 
I pushed a fix to this pull request. It's only deleting 3 lines of code in module_wind_fitch.F (lines 206, 209 and 214 in the current version in the develop branch). 

Could you update before release?
Thanks, Balthazar

dudhia commented:
You would need to send it as a new pull request for develop. We still can fix it.

SengersB commented:
Understood, I'll do that tomorrow. 
Thanks for the quick reply!
